### ISSUE TITLE
What's the difference between "RegL1Loss" and "RegWeightedL1Loss"

### ISSUE THREAD

**Issue by seekFire** [xingyizhou/CenterNet#401]
In "RegL1Loss", there has the code as follows:
    mask = mask.unsqueeze(2).expand_as(pred).float()

while in "RegWeightedL1Loss", there is only:
    mask = mask.float()

However, due to the broadcast operation, the calculation of them will have the same results:
    loss = F.l1_loss(pred * mask, target * mask, size_average=False)

So where should be different between those two loss functions?

**Comment by jhultman** [comment]
I haven't looked too closely at `losses.py` but I imagine the difference is in the [denominator](https://github.com/xingyizhou/CenterNet/blob/9efed05d531922b001b101c648c7ac5af3a9dafb/src/lib/models/losses.py#L148). `mask.sum()` will be larger for the explicitly expanded mask:

```
import torch
mask = torch.ones((1))
mask.sum()
>> 1
mask.unsqueeze(1).expand(1, 3).sum()
>> 3
```

**Comment by xingyizhou** [comment]
Hi,
  Sorry for the confusion. @jhultman is correct. RegWeightedL1Loss takes an additional weight mask. It is used in human pose estimation when some joints are invisible/ out of the image.

**Comment by seekFire** [comment]
Oh, I get it! Thank you very much! @jhultman @xingyizhou

**Comment by xingyizhou** [comment]
@zhiqwang I didn't understand your question.

**Comment by zhiqwang** [comment]
@xingyizhou Thanks for your reply. In `CtdetLoss`, the `self.crit_reg` can be chose between `RegL1Loss` and `RegLoss`. So the logic of this two loss is same?
https://github.com/xingyizhou/CenterNet/blob/9efed05d531922b001b101c648c7ac5af3a9dafb/src/lib/trains/ctdet.py#L21-L22

But I noticed the denominator in this two loss is different. In `RegLoss`, the denominator is computed before the `expand_as` function. As in `RegL1Loss`, the denominator is computed after the `expand_as` function.
https://github.com/xingyizhou/CenterNet/blob/9efed05d531922b001b101c648c7ac5af3a9dafb/src/lib/models/losses.py#L143-L149

**Comment by xingyizhou** [comment]
Hi @zhiqwang . Thanks for the detailed reading. These two are different. As far as I recall, the other one is from the [3DTracking repo](https://github.com/ucbdrive/3d-vehicle-tracking) for our experiments on KITTI. I only have tried the current version for ctdet. Later after publishing the code I have tried using the ctdet regression loss for KITTI as well. The results are similar. I would say always using the current version for ctdet is safe.

**Comment by zhiqwang** [comment]
@xingyizhou Thanks for your reply and awesome work. I'll do more experiments based on your work.

**Comment by ujsyehao** [comment]
@zhiqwang, You can refer to https://github.com/xingyizhou/CenterNet/issues/647
